rfcs/text/0011_global_search.md

```diff
@@ -0,0 +1,693 @@
+  exactly the same.
+- `coreStart.globalService` exposes an API to be able to search for objects. This API is available from both public
+  and server sides. 
+  - When using the server `find` API, only results from providers registered from the server will be returned. 
+  - When using the public `find` API, results from provider registered from either server and public sides will be returned.
+- During a `find` call, the service will call all the registered result providers and collect their result observables.
+  Every time a result provider emits some new results, the `globalSearch` service will:
+  - Consolidate/enhance them
+  - Merge them with the already present results
+  - Sort and order the new aggregated results
+  - Emit this up to date list of results
+
+### result provider registration
+
+Due to the fact that some kind of results (i.e `application`, and maybe later `management_section`) only exists on 
+the public side of Kibana and therefor are not known on the server side, the `registerResultProvider` API will be 
+available both from the public and the server counterpart of the `GlobalSearchService`.
+
+However, as results from providers registered from the client-side will not be available from the server's `find` API, 
+registering result providers from the client should only be done to answer this specific use case and will be 
+discouraged, by providing appropriated jsdoc and documentation explaining that it should only
+be used when it is not technically possible to register it from the server side instead.
+
+### results url consolidation
+
+When retrieving results from providers, the GS service will always start by consolidating them. The most notable (and
+currently only) step is to convert the result url to an absolute one.
+
+#### absolute url conversion logic
+
+Results returned from the GlobalSearch's `find` programmatic and HTTP APIs will all contains absolute urls for the following
+reasons:
+- It would not be usable by external consumers otherwise. 
+- Some result providers are supposed to be returning results from outside of the current `basePath`, or even from outside of the
+  Kibana instance (ie: results from another space, results from another Kibana instance in a cloud cluster)
+
+However, as forging absolute urls can be a tedious process for the plugins, the `url` property of results returned by
+a result provider can be either an absolute url or an url path relative to the executing request's `basePath`.
+
+I.E are considered valid:
+- `https://my-other-kibana-instance/some/result`
+- `/app/kibana#/dashboard/some-id`
+
+When consolidating the result, the logic regarding the `url` property is:
+- if the `url` is absolute, return it unchanged
+- if the `url` is a relative path (starts with `/`), it will be converted to an absolute url by prepending the Kibana
+instance's newly introduced `publicAddress`.
+
+#### server.publicAddress
```

Comment:
It's incredibly common for Kibana to be hosted behind a reverse-proxy. We have similar settings for specific use-cases, and they've honestly been nothing but trouble. Reporting has `xpack.reporting.kibanaServer.*` and it's been a constant source of SDHs ever since it was implemented. Security has `xpack.security.public.*` (now deprecated) which caused so many issues we got rid of it.

We're stuck deciding between returning relative URLs, which some consumers must then translate to absolute URLs; or adding this setting, which likely leads to us returning incorrect URLs. I'd much prefer that we return relative URLs, and require the consumer to translate them into absolute URLs. When Kibana is consuming these URLs client-side, we don't even need to do this translation. If we do happen to have external consumers of the HTTP APIs, they're already going to have to know about the public host/port/protocol of Kibana to be able to use the HTTP API that requiring them to do this translation seems reasonable.

Reply:
Yea, this keeps going back in the discussion. 

Main issue with relative urls is that they are relative to the current request's basePath. For instance when returning results for a request within a space, when a provider returns `/dashboard/my-dashboard`, this is not the 'absolute' path, which would be `/my-space/dashboard/my-dashboard` (or even `/kibana-base-path/space/dashboard/my-dashboard` if kibana has a basepath defined).

Should we then return the path prepended by the current's request whole basePath? only the 'extra' path (aka 'requestScopePath' in `src/core/server/http/base_path_service.ts`), or not prepended at all?

Also, we have to keep in mind that providers will be able to return urls from other kibana instances, meaning that we will have to support absolute urls handling on the client-side anyway. Not converting the 'local' urls to absolute one seems like introducing more inconsistency and therefor complexity.

One option would be to return relative urls for the `internal` API, and put this discussion back on the table when we actually have an external provider. However I'm with Josh on this one: I don't see how an external consumer will be able to properly handle directing to a result if the urls are not absolute.

wdyt?

Reply:
> The main use case we're considering here is ChatOps which will not know the public host/port/protocol of Kibana since it is running in the same process but will be publishing links externally.

Perhaps I'm misunderstanding some behavior of ChatOps or overlooking some details. ChatOps will require that custom code be written to integrate with these third-party services, like Slack. This custom code would need to know that Kibana is hosted at some protocol/host/port/base-path to be able to call the Kibana APIs that make ChatOps possible. Couldn't this custom code then translate the response from the Kibana APIs which include relative URLs into full URLs which includes the public protocol/host/port/base-path?

Is there some use-case for ChatOps to publish these URLs to some other destination that doesn't know where Kibana is hosted?

> Should we then return the path prepended by the current's request whole basePath? only the 'extra' path (aka 'requestScopePath' in src/core/server/http/base_path_service.ts), or not prepended at all?

I was originally thinking we'd return URLs prepended by the current request's whole basePath, for example `/foo/s/marketing/app/dashboards`. If this string was put directly into a link's `href` property, the user would be redirected properly; or if this URL was used with a XMLHTTPRequest, it'd work as well.

> Also, we have to keep in mind that providers will be able to return urls from other kibana instances, meaning that we will have to support absolute urls handling on the client-side anyway. Not converting the 'local' urls to absolute one seems like introducing more inconsistency and therefor complexity.

I'm not really following. When used in the context of the browser, we don't need to translate local urls into absolute opens. Assuming that Kibana is hosted at `http://some-host/foo` the following two usages are equivalent:

```
const xhr1 = new XMLHttpRequest();
xhr1.open('GET', 'http://some-host/foo/s/marketing/app/dashboards');

const xhr2 = new XMLHttpRequest();
xhr2.open('GET', '/foo/s/marketing/app/dashboards');
```

Reply:
> I'm not really following. When used in the context of the browser, we don't need to translate local urls into absolute opens

True. What I meant was that to perform SPA navigation to a given (local) result, we will need to parse the url of the result to check if it belongs to the current kibana instance's domain or not to choose if we will be using `application.navigateTo` or just `location.assign`. Having relative url would introduce more complexity to the parsing, as we would need to handle both absolute and relative urls.

One compromise would be to only parse relative urls and assume that absolute ones would always be using `location.assign`. Drawback is that a provider MUST return a relative url when referencing a result for current kibana instance. But that seems acceptable I guess.

Reply:
> One compromise would be to only parse relative urls and assume that absolute ones would always be using location.assign. Drawback is that a provider MUST return a relative url when referencing a result for current kibana instance. But that seems acceptable I guess.

This seems like a reasonable compromise to me.

Reply:
After thinking about it again, and after actually having implemented a very similar url parsing mechanism (in https://github.com/elastic/kibana/pull/65164/files#diff-ee8cabd7d05b6fdac27fa3d6093ea4fc), I think that for the internal API, we can allow mixed content of absolute url and relative paths (including the basePath).

Parsing both would be possible, meaning that the `result.navigate()` would be able to perform SPA navigation even with absolute urls pointing to the instance's origin (even if it's very unlikely that 'local' result providers would return absolute urls instead of relative one)

Major benefit is avoiding the introduction of this complicated and troublesome `publicAddress` feature/property. We'll see later, once the need for a public API/endpoint arises, if we want/need this for the public endpoint

Reply:
`publicAddress` proposal removed in https://github.com/elastic/kibana/pull/64284/commits/aed714fd8664ae56ae9e46878e4bdd2c6f05ec23